File: Basic_Bit_Operations/Q2. Single Number/Q2_SingleNumber.java
class Solution 
{
    public int singleNumber(int[] nums) 
    {
        int result = 0;

        // loop through all numbers
        for (int num : nums) 
        {
            // XOR with result: duplicates cancel out
            result ^= num; 
        }

        // the remaining number is the one that appears only once
        return result;
    }
}

File: Basic_Bit_Operations/Q2. Single Number/Q2_SingleNumber.py
class Solution(object):
    def singleNumber(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """

        result = 0

        # loop through all numbers
        for num in nums:
            # XOR cancels out duplicates
            result ^= num

        # the remaining number is the one that appears only once
        return result
